### ISSUE TITLE
Question about resource status after different failed restart scenarios

### ISSUE THREAD

**Issue by blogh** [dalibo/PAF#25]
Hi,

I tested the following scenario. I am new to pacemaker so i am not sure the result is intended (and if the test is worth something), i am also not 100% sure if it's tied to the agent or a on-fail clause of the configuration. Sorry for the disturbance if it's normal or unrelated to the agent.

I use:
- Ubuntu 14.04
- Pacemaker 1.1.10 / Corosync 2.3.3
- Paf 1.0.2
- PostgreSQL 9.4

Cluster in nominal state:

```
 Resource Group: master-group
     vip-master (ocf::heartbeat:IPaddr2):       Started x64lmwbio9f-priv
     vip-rep    (ocf::heartbeat:IPaddr2):       Started x64lmwbio9f-priv
 Master/Slave Set: msPostgresql [pgsqlsrv]
     Masters: [ x64lmwbio9f-priv ]
     Slaves: [ x64lmwbio9g-priv ]
 Clone Set: fencing [st-null]
     Started: [ x64lmwbio9f-priv x64lmwbio9g-priv ]
```

Stop the node x64lmwbio9g-priv  with "sudo service pacemaker stop"
Modify the postgresql.conf to introduce an error during startup.
Start the node x64lmwbio9g-priv  with "sudo service pacemaker start"

The cluster ends up like this. (i wasn't surprised)

```
 Resource Group: master-group
     vip-master (ocf::heartbeat:IPaddr2):       Started x64lmwbio9f-priv
     vip-rep    (ocf::heartbeat:IPaddr2):       Started x64lmwbio9f-priv
 Master/Slave Set: msPostgresql [pgsqlsrv]
     Masters: [ x64lmwbio9f-priv ]
     Stopped: [ x64lmwbio9g-priv ]
 Clone Set: fencing [st-null]
     Started: [ x64lmwbio9f-priv x64lmwbio9g-priv ]
```

Now if i do the following:

Stop the node x64lmwbio9g-priv  with "sudo service pacemaker stop"
Modify the recovery.conf.pcmk application_name (pgsql_validate_all fails) (It happened once when I forgot to restore the configuration after a hasty pg_rewind) 
Start the node x64lmwbio9g-priv  with "sudo service pacemaker start"

Full list of resources:

```
 Resource Group: master-group
     vip-master (ocf::heartbeat:IPaddr2):       Stopped
     vip-rep    (ocf::heartbeat:IPaddr2):       Stopped
 Master/Slave Set: msPostgresql [pgsqlsrv]
     Stopped: [ x64lmwbio9f-priv x64lmwbio9g-priv ]
 Clone Set: fencing [st-null]
     Started: [ x64lmwbio9f-priv x64lmwbio9g-priv ]
```

I understand that application_name is a prerequisite for the agent to work but i am surprised that a wrong configuration on one node compromises the whole cluster. 

I can provide more info if needed. 

Benoit.

Side note: with 1.0.0 the vip- group didn't stop in the last case, with 1.0.2 it works as intended which is cool !

**Comment by blogh** [comment]
After reading a little about the doc [1]. What about returning : OCF_ERR_ARGS or OCF_ERR_INSTALLED instead of OCF_ERR_CONFIGURED. (to avoid the fatal return code)

[1] http://clusterlabs.org/doc/en-US/Pacemaker/1.1-crmsh/html/Pacemaker_Explained/s-ocf-return-codes.html

**Comment by ioguix** [comment]
Hello @blogh 

I can not imagine a scenario in your second test that would lead to this situation. A look at the log files would probably help though.

Concerning the return code, OCF_ERR_ARGS or OCF_ERR_INSTALLED are not accurate with the situation you are testing. Having a bad recovery.conf is a configuration issue. However, the only consequence should be a stop or fence of the resource on this node...

**Comment by blogh** [comment]
Hi,

The documentation  [1] says that :

- fatal : Stop the resource and prevent it from being started on any cluster node 
- hard: Move the resource elsewhere and prevent it from being retried on the current node.

I was under the impression that hard would be better.  In my case, I just misconfigured the standby while reintroducing it into the cluster (after pg_rewind). I thought that the ressource would be stopped only on that node but the primary becase unavailable too.

I must have misunderstood something.

I will run the test again and provide configuration and log details.
I might have done something dumb :)

Benoit.

[1] http://clusterlabs.org/doc/en-US/Pacemaker/1.1-crmsh/html/Pacemaker_Explained/_how_are_ocf_return_codes_interpreted.html

**Comment by blogh** [comment]
Hi, 

you will find attached two testcases :

- application_name is wrong, but the node name is a substring of application name.  (config_mistake_negativescore*)

In this case we fail in _check_locations and the ressource is in slave state with negative score.

```
        if ( $partition_nodes !~ /$row->[0]/ ) {
            ocf_log( 'info', sprintf
                '_check_locations: ignoring unknown application_name/node "%s"', $row->[0] );
            next;
        }
```


- application_name is wrong (config_mistake_stop*)

In this case we fail in pgsql_validate_all and the resource is stopped on all nodes. 

```
    unless ( grep /^\s*primary_conninfo\s*=.*application_name=$nodename/, @content ) {
        ocf_log( 'crit',  sprintf
            "Recovery template file must contain in primary_conninfo parameter \"application_name=%s\"",
            $nodename );
        exit $OCF_ERR_CONFIGURED;
    }
```


In my opinion both are incorrect, the ressource should be stopped only on the node with the wrong configuration. 

I can provide mode info if needed.

Benoit.

(config_mistake_negativescore
[config_mistake_negativescore_x64lmwbio9f.txt](https://github.com/dalibo/PAF/files/302258/config_mistake_negativescore_x64lmwbio9f.txt)
[config_mistake_negativescore_x64lmwbio9g.txt](https://github.com/dalibo/PAF/files/302260/config_mistake_negativescore_x64lmwbio9g.txt)
[config_mistake_stop.txt](https://github.com/dalibo/PAF/files/302262/config_mistake_stop.txt)
[config_mistake_stop_x64lmwbio9f.txt](https://github.com/dalibo/PAF/files/302261/config_mistake_stop_x64lmwbio9f.txt)
[config_mistake_stop_x64lmwbio9g.txt](https://github.com/dalibo/PAF/files/302263/config_mistake_stop_x64lmwbio9g.txt)
[config_mistake_negativescore.txt](https://github.com/dalibo/PAF/files/302259/config_mistake_negativescore.txt)
[pacemaker_conf.txt](https://github.com/dalibo/PAF/files/302282/pacemaker_conf.txt)

**Comment by ioguix** [comment]
Hi,

Good catch for the first scenario, having the hostname as a substring in the application_name. The expected behavior should be the same than in your second scenario: pgsql_vaidate_all should catch the error (the negative score is right though).This should definitely be fixed. Thank you for this report.

I might missing something about the second scenario. During my tests, I don't remember a fatal error on a slave would take down the whole cluster. The only difference I can see is that I do my tests on a three node cluster with quorum enabled (and stonith, but you have it as well). I understand a fatal error would forbid the resource to run anywhere in the cluster, but I thought this was related to the local cloned resource in the ms resource, not the ms resource itself.

I need some more time to investigate this. I'm on the fence between enforcing clean setup on the whole cluster and keep things "simple" (from the code point of view) or trying to decide when we can isolate a local error and return a hard error. But I'm kind of afraid of opening a can of worm here :/

**Comment by ioguix** [comment]
This has been fixed in e30eb063d17e2ded6f51b9ae7b6a4a33b7d606ff
